moshi/src/test/java/com/squareup/moshi/JsonAdapterTest.java

```diff
@@ -182,4 +183,36 @@
     serializeNulls.toJson(writer, Collections.<String, String>singletonMap("a", null));
     assertThat(factory.json()).isEqualTo("{\"a\":null}");
   }
+
+  @Test public void stringDocumentMustBeFullyConsumed() throws IOException {
+    JsonAdapter<String> brokenAdapter = new JsonAdapter<String>() {
+      @Override public String fromJson(JsonReader reader) throws IOException {
+        return "Forgot to call reader.nextString().";
+      }
+
+      @Override public void toJson(JsonWriter writer, @Nullable String value) throws IOException {
+        throw new AssertionError();
+      }
+    };
+    try {
+      brokenAdapter.fromJson("\"value\"");
+      fail();
+    } catch (JsonDataException e) {
+      assertThat(e).hasMessage("JSON document was not fully consumed.");
+    }
+  }
+
+  @Test public void lenientAdapterDoesNotCheckStringDocumentConsumption() throws IOException {
+    JsonAdapter<String> brokenAdapter = new JsonAdapter<String>() {
+      @Override public String fromJson(JsonReader reader) throws IOException {
+        return "Forgot to call reader.nextString().";
+      }
+
+      @Override public void toJson(JsonWriter writer, @Nullable String value) throws IOException {
+        throw new AssertionError();
+      }
+    }.lenient();
```

Comment:
I think this is broken. Try lenient().nullSafe().

Reply:
ahaha, thanks for the check. i'm going to try my original thought and make an overridable implementation of fromJson(String).

Reply:
Maybe just a package-private method, `isLenient()` that returns false by default, returns true for lenient adapters, and delegates for other wrappers?

Reply:
yep, i don't see a way to let the wrappers override fromJson(String) correctly.
isLenient is easier.